File: Source/Aladdin.CAPI.Java/src/aladdin/capi/software/ConfigStore.java
package aladdin.capi.software;
import aladdin.*;
import aladdin.capi.*; 
import aladdin.io.xml.*; 
import java.io.*;
import java.util.*; 

///////////////////////////////////////////////////////////////////////////
// Раздел файла конфигурации как хранилище программных контейнеров
///////////////////////////////////////////////////////////////////////////
public class ConfigStore extends ContainerStore
{
	// модель документа
	private XmlKey document; 

	// конструктор
	public ConfigStore(CryptoProvider provider, Scope scope, String configName) 
    {
        // сохранить переданные параметры
        super(provider, scope); document = null; 
        
        // получить профиль пользователя
        String directory = scope.equals(Scope.SYSTEM) ? 
            OS.INSTANCE.getSharedFolder() : OS.INSTANCE.getUserFolder(); 

        // указать полный путь к файлу
        String configFile = String.format(
            "%1$s%2$s%3$s", directory, File.separator, configName
        ); 
        try {
            // при отсутствии файла
            File file = new File(configFile); if (!file.exists()) 
            {
                // создать документ
                document = XmlKey.createDocument(configFile, "configuration"); 
            }
            // открыть документ
            else document = XmlKey.openDocument(configFile, "rw"); 
        }
        // обработать возможное исключение
        catch (Throwable e) {}
    }
    // имя хранилища
    @Override public Object name() 
    {
        // имя хранилища
        return scope().equals(Scope.SYSTEM) ? "FSLM" : "FSCU"; 
    }
    ///////////////////////////////////////////////////////////////////////
	// Управление контейнерами
	///////////////////////////////////////////////////////////////////////
	@Override public String[] enumerateObjects()
	{
        // проверить поддержку операции
        if (document == null) return new String[0]; 
        
        // создать список имен
        List<String> names = new ArrayList<String>(); 

        // получить элемент для контейнеров
        XmlKey key = document.openKey("containers", "r"); 
             
        // проверить наличие элемента
        if (key == null) return new String[0]; 

        // для всех каталогов
        for (XmlKey child : key.enumerateKeys("r"))
        {
            // добавить каталог в список
            names.add(child.getAttribute("name", null));
        }
	    // вернуть список имен
        return names.toArray(new String[names.size()]); 
	}
	///////////////////////////////////////////////////////////////////////
	// Управление физическими потоками
	///////////////////////////////////////////////////////////////////////
    @Override protected ContainerStream createStream(Object name) throws IOException
    {
        // проверить поддержку операции
        if (document == null) throw new UnsupportedOperationException(); 
        
        // получить элемент для контейнеров
        XmlKey key = document.openOrCreateKey("containers"); 
             
        // для всех контейнеров
        for (XmlKey child : key.enumerateKeys("r"))
        {
            // получить имя контейнера
            String value = child.getAttribute("name", null); 

            // сравнить имена контейнеров
            if (value.compareToIgnoreCase(name.toString()) == 0) 
            { 
                // проверить отсутствие элемента
                throw new IOException(); 
            }
        }
        // добавить новый элемент 
        XmlKey created = document.createKey("container"); 
            
        // указать имя контейнера
        created.setAttribute("name", name.toString()); 
            
        // вернуть поток данных в файле конфигурации
        return new ElementStream(created);
    }
    @Override protected ContainerStream openStream(Object name, String access) throws IOException
    {
        // проверить поддержку операции
        if (document == null) throw new UnsupportedOperationException(); 
        
        // получить элемент для контейнеров
        XmlKey key = document.openKey("containers", access); 

        // проверить наличие элемента
        if (key == null) throw new NoSuchElementException(); 

        // для всех контейнеров
        for (XmlKey child : key.enumerateKeys(access))
        {
            // получить имя контейнера
            String value = child.getAttribute("name", null); 

            // сравнить имена контейнеров
            if (value.compareToIgnoreCase(name.toString()) == 0) 
            { 
                // вернуть поток данных в файле конфигурации
                return new ElementStream(child); 
            }
        }
        // выбросить исключение
        throw new NoSuchElementException(); 
    }
    @Override protected void deleteStream(Object name) throws IOException
    {
        // проверить поддержку операции
        if (document == null) throw new UnsupportedOperationException(); 

        // получить элемент для контейнеров
        XmlKey key = document.openKey("containers", "rw"); 
             
        // проверить наличие элемента
        if (key == null) return; 

        // для всех контейнеров
        for (XmlKey child : key.enumerateKeys("r"))
        {
            // получить имя контейнера
            String value = child.getAttribute("name", null); 

            // сравнить имена контейнеров
            if (value.compareToIgnoreCase(name.toString()) == 0) 
            { 
                // удалить контейнер
                key.deleteKey(child); return; 
            }
        }
    }
	///////////////////////////////////////////////////////////////////////////
    // Поток хранилища данных в файле конфигурации
	///////////////////////////////////////////////////////////////////////////
	private static class ElementStream extends ContainerStream
    {
	    // конструктор
	    public ElementStream(XmlKey element) 
	    
            // сохранить переданные параметры
            { this.element = element; } private final XmlKey element; 
        
        // имя контейнера
        @Override public Object name() { return uniqueID(); }
        
        // уникальный идентификатор
        @Override public String uniqueID() { return element.getAttribute("name", null); }
        
        // прочитать данные
        @Override public byte[] read() throws IOException
        {
            // получить содержимое элемента
            return Base64.getDecoder().decode(element.getValue()); 
        }
        // записать данные
		@Override public void write(byte[] buffer) throws IOException
		{
            // записать данные
            element.setValue(Base64.getEncoder().encodeToString(buffer)); 
		}
	}
}

File: Source/Aladdin.CAPI/CAPI/Software/ConfigStore.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace Aladdin.CAPI.Software
{
	///////////////////////////////////////////////////////////////////////////
	// Раздел файла конфигурации как хранилище программных контейнеров
	///////////////////////////////////////////////////////////////////////////
	public class ConfigStore : ContainerStore
	{
		// модель документа
		private IO.Xml.XmlKey document; 

		// конструктор
		public ConfigStore(CryptoProvider provider, Scope scope, string configName) 
            
            // сохранить переданные параметры
            : base(provider, scope) 
        {
            // указать тип каталога
            Environment.SpecialFolder type = (scope == Scope.System) ? 
                Environment.SpecialFolder.CommonApplicationData : 
                Environment.SpecialFolder.ApplicationData; 

            // получить профиль пользователя
            string directory = Environment.GetFolderPath(type); 

            // указать полный путь к файлу
            string configFile = String.Format("{0}{1}{2}", 
                directory, Path.DirectorySeparatorChar, configName
            ); 
            try {
                // при отсутствии файла
                if (!File.Exists(configFile))
                {
                    // создать документ
                    document = IO.Xml.XmlKey.CreateDocument(
                        configFile, "configuration"
                    ); 
                }
                else {
                    // открыть документ
                    document = IO.Xml.XmlKey.OpenDocument(
                        configFile, FileAccess.ReadWrite
                    ); 
                }
            }
            catch {} 
		}  
        // имя хранилища
        public override object Name
        {
            // имя хранилища
            get { return (Scope == Scope.System) ? "FSLM" : "FSCU"; }
        }
		///////////////////////////////////////////////////////////////////////
		// Управление контейнерами
		///////////////////////////////////////////////////////////////////////
		public override string[] EnumerateObjects()
		{
            // проверить поддержку операции
            if (document == null) return new string[0]; 

            // создать список имен
            List<String> names = new List<String>(); 

            // получить элемент для контейнеров
            IO.Xml.XmlKey key = document.OpenKey("containers", FileAccess.Read); 
             
            // проверить наличие элемента
            if (key == null) return new string[0]; 

            // для всех каталогов
            foreach (IO.Xml.XmlKey child in key.EnumerateKeys(FileAccess.Read))
            {
                // добавить каталог в список
                names.Add(child.GetAttribute("name", null));
            }
	        // вернуть список имен
            return names.ToArray(); 
		}
		///////////////////////////////////////////////////////////////////////
		// Управление физическими потоками
		///////////////////////////////////////////////////////////////////////
        protected override ContainerStream CreateStream(object name)
        {
            // проверить поддержку операции
            if (document == null) throw new NotSupportedException(); 

            // получить элемент для контейнеров
            IO.Xml.XmlKey key = document.OpenOrCreateKey("containers"); 
             
            // для всех контейнеров
            foreach (IO.Xml.XmlKey child in key.EnumerateKeys(FileAccess.Read))
            {
                // получить имя контейнера
                string value = child.GetAttribute("name", null); 

                // сравнить имена контейнеров
                if (String.Compare(value, name.ToString(), true) == 0) 
                { 
                    // проверить отсутствие элемента
                    throw new IOException(); 
                }
            }
            // добавить новый элемент 
            IO.Xml.XmlKey created = document.CreateKey("container"); 
            
            // указать имя контейнера
            created.SetAttribute("name", name.ToString()); 
            
            // вернуть поток данных в файле конфигурации
            return new ElementStream(created);
        }
        protected override ContainerStream OpenStream(object name, FileAccess access)
        {
            // проверить поддержку операции
            if (document == null) throw new NotSupportedException(); 

            // получить элемент для контейнеров
            IO.Xml.XmlKey key = document.OpenKey("containers", access); 

            // проверить наличие элемента
            if (key == null) throw new NotFoundException(); 

            // для всех контейнеров
            foreach (IO.Xml.XmlKey child in key.EnumerateKeys(access))
            {
                // получить имя контейнера
                string value = child.GetAttribute("name", null); 

                // сравнить имена контейнеров
                if (String.Compare(value, (string)name, true) == 0) 
                { 
                    // вернуть поток данных в файле конфигурации
                    return new ElementStream(child); 
                }
            }
            // контейнер не найден
            throw new NotFoundException(); 
        }
        protected override void DeleteStream(object name)
        {
            // проверить поддержку операции
            if (document == null) throw new NotSupportedException(); 

            // получить элемент для контейнеров
            IO.Xml.XmlKey key = document.OpenKey("containers", FileAccess.ReadWrite); 
             
            // проверить наличие элемента
            if (key == null) return; 

            // для всех контейнеров
            foreach (IO.Xml.XmlKey child in key.EnumerateKeys(FileAccess.Read))
            {
                // получить имя контейнера
                string value = child.GetAttribute("name", null); 

                // сравнить имена контейнеров
                if (String.Compare(value, (string)name, true) == 0) 
                { 
                    // удалить контейнер
                    key.DeleteKey(child); return; 
                }
            }
        }
	    ///////////////////////////////////////////////////////////////////////////
        // Поток хранилища данных в файле конфигурации (только для чтения)
	    ///////////////////////////////////////////////////////////////////////////
	    private class ElementStream : ContainerStream
        {
            // узел элемента
            private IO.Xml.XmlKey element; 

	        // конструктор
	        public ElementStream(IO.Xml.XmlKey element)
	        {
                // сохранить переданные параметры
                this.element = element; 
            }
            // имя контейнера
            public override object Name { get { return UniqueID; }}
        
            // уникальный идентификатор
            public override string UniqueID 
            { 
                // уникальный идентификатор
                get { return element.GetAttribute("name", null); }
            }
            // прочитать данные
            public override byte[] Read()
            {
                // раскодировать содержимое элемента
                return Base64.GetDecoder().Decode(element.GetValue()); 
            }
            // записать данные
		    public override void Write(byte[] buffer)
		    {
                // записать данные
                element.SetValue(Base64.GetEncoder().EncodeToString(buffer)); 
		    }
	    }
	}
}
